### PR TITLE
Refactor wicked tests 

### PR DESCRIPTION
**Pull request by asmorodskyi**
- Related ticket: https://progress.opensuse.org/issues/32959
- Verification run: 
http://kimball.arch.suse.de/tests/1389
http://kimball.arch.suse.de/tests/1388

### CODE REVIEW

File: tests/wicked/advanced_sut.pm [hunk 1]
```diff
@@ -40,8 +42,20 @@ sub run {
 
     # Create mutex to unlock REF
     mutex_create("test_1_ready");
+    if (!$ret) {
+        record_info("Can't ping IP $parallel_ip_in_tunnel", result => 'fail');
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: We could add a pre-string to this message to identify easily in the logs what failed. Example:
record_info("[Test-1] Can't ping IP $parallel_ip_in_tunnel", result => 'fail');

File: tests/wicked/advanced_sut.pm [hunk 2]
```diff
@@ -23,6 +24,7 @@ use mmapi;
 
 sub run {
     my ($self) = @_;
+    my %results;
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: +1 for this variable to store results of sub-tests

**Comment by asmorodskyi** [comment]
@okurz curious  what you think about such approach

File: tests/wicked/advanced_sut.pm (hunk 1 above)
Comment (asmorodskyi) [reply tests/wicked/advanced_sut.pm]: will do that

**Comment by asmorodskyi** [comment]
@jlausuch  this is how it looks with your PR merged

File: tests/wicked/advanced_sut.pm [hunk 3]
```diff
@@ -21,11 +22,28 @@ use utils 'systemctl';
...
+
 sub run {
     my ($self) = @_;
-    record_info('Test 1', 'Create a gre interface from legacy ifcfg files');
+    my %results;
+    my $iface      = script_output('ls /sys/class/net/ | grep -v lo | head -1');
     my $gre_config = '/etc/sysconfig/network/ifcfg-gre1';
+    my $sit_config = '/etc/sysconfig/network/ifcfg-sit1';
     $self->get_from_data('wicked/ifcfg-gre1_', $gre_config, add_suffix => 1);
+    $self->get_from_data('wicked/ifcfg-sit1_', $sit_config, add_suffix => 1);
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: if we copy this file without replacing the dummy values ('local_ip', 'remote_ip', 'tunnel_ip') and we do an ifup, wicked will complain. So, I would move this line to the corresponding test case block.

See this: http://fromm.arch.suse.de/tests/677/file/serial_terminal.txt
    ifup gre1 ; echo RvmM0 -$?-
    wicked: ifcfg-sit1: Cannot parse TUNNEL_LOCAL_IPADDR="local_ip"
    gre1            up

File: tests/wicked/advanced_sut.pm [hunk 4]
```diff
@@ -36,16 +54,13 @@ sub run {
...
-    record_info("[Test 1] Can't ping IP $parallel_ip_in_tunnel") if !$ret;
-    #TODO: Reset network (delete gre1 interface)
 
-    record_info('Test 3', 'Create a SIT interface from legacy ifcfg files');
-    my $sit_config = '/etc/sysconfig/network/ifcfg-sit1';
-    $self->get_from_data('wicked/ifcfg-sit1_', $sit_config, add_suffix => 1);
+    #$self->before_scenario('Test 2', 'Create a gre interface from wicked XML files');
+
+    $self->before_scenario('Test 3', 'Create a SIT interface from legacy ifcfg files', $iface);
     my $ip_in_tunnel          = $self->get_ip(is_wicked_ref => 0, type => 'sit_tunnel_ip');
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: remove "my" here, as the variables are initialized above

File: tests/wicked/advanced_sut.pm [hunk 5]
```diff
@@ -36,29 +53,33 @@ sub run {
     assert_script_run("cat $gre_config");
     assert_script_run('ifup gre1');
     assert_script_run('ip a');
-    my $ret = $self->ping_with_timeout(ip => "$parallel_ip_in_tunnel", timeout => '60');
+    $results{Test1} = get_test_result($ip_in_tunnel, "");
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: $results{Test1} = get_test_result($self, $ip_in_tunnel, "");

File: tests/wicked/advanced_sut.pm [hunk 6]
```diff
@@ -36,29 +53,33 @@ sub run {
...
+    $ip_in_tunnel          = $self->get_ip(is_wicked_ref => 0, type => 'sit_tunnel_ip');
+    $parallel_ip_in_tunnel = $self->get_ip(is_wicked_ref => 1, type => 'sit_tunnel_ip');
     assert_script_run("sed \'s/local_ip/$ip_no_mask/\' -i $sit_config");
     assert_script_run("sed \'s/remote_ip/$parallel_host_ip_no_mask/\' -i $sit_config");
     assert_script_run("sed \'s/tunnel_ip/$ip_in_tunnel\\/127/\' -i $sit_config");
     assert_script_run("cat $sit_config");
     assert_script_run('ifup sit1');
     assert_script_run('ip a');
-    $ret = $self->ping_with_timeout(ip => "$parallel_ip_in_tunnel", timeout => '60', ip_version => 'v6');
+    $results{Test3} = get_test_result($ip_in_tunnel, "v6");
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: $results{Test3} = get_test_result($self, $ip_in_tunnel, "v6");

File: tests/wicked/advanced_sut.pm [hunk 7]
```diff
@@ -21,11 +22,27 @@ use utils 'systemctl';
...
+    }
+}
+
 sub run {
     my ($self) = @_;
-    record_info('Test 1', 'Create a gre interface from legacy ifcfg files');
+    my %results;
+    my $iface      = script_output('ls /sys/class/net/ | grep -v lo | head -1');
     my $gre_config = '/etc/sysconfig/network/ifcfg-gre1';
+    my $sit_config = '/etc/sysconfig/network/ifcfg-sit1';
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: I would also move this line to the corresponding test. If we keep all the things that correspond to a sub-test in the sub-test block, we can think of a smart way of grouping things together to not write similar lines again and again.

File: tests/wicked/advanced_sut.pm (hunk 5 above)
Comment (jlausuch) [reply tests/wicked/advanced_sut.pm]: not important  now, but shall we keep the keys in lower case? i.e. $results{test1}, $results{test2}... or just numbers, $results{1}, $results{2}, 
This way, we can always numbers to some text and be more flexible when presenting the results.

File: tests/wicked/advanced_sut.pm [hunk 8]
```diff
@@ -10,6 +10,7 @@
 # Summary: Advanced test cases for wicked
 # Test scenarios:
 # Test 1 : Create a GRE interface from legacy ifcfg files
+# Test 2 : Create a gre interface from wicked XML files
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: Add
# Test 3 : Create a SIT interface from legacy ifcfg files

File: tests/wicked/advanced_sut.pm [hunk 9]
```diff
@@ -10,6 +10,7 @@
...
 # Maintainers:
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: Travis complains of this line. Missing '# Maintainer: <email address>' in tests/wicked/advanced_sut.pm

File: tests/wicked/advanced_sut.pm [hunk 10]
```diff
@@ -21,44 +19,64 @@ use utils 'systemctl';
...
+        record_info("Can't ping IP $ip", result => 'fail');
+        return "FAILED";
+    }
+    else {
+        return "PASSED";
+    }
+}
+
+sub setup_tunnel {
+    my ($self, @args) = @_;
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: %args

File: tests/wicked/advanced_sut.pm [hunk 11]
```diff
@@ -21,44 +19,64 @@ use utils 'systemctl';
...
-    my $ret = $self->ping_with_timeout(ip => "$parallel_ip_in_tunnel", timeout => '60');
+    my %results;
+    my $iface  = script_output('ls /sys/class/net/ | grep -v lo | head -1');
+    my $config = '/etc/sysconfig/network/ifcfg-gre1';
+    $self->get_from_data('wicked/ifcfg-gre1_', $config, add_suffix => 1);
+    $self->before_scenario('Test 1', 'Create a gre interface from legacy ifcfg files');
+    my $tunnel_ip        = $self->get_ip(is_wicked_ref => 0, type => 'gre_tunnel_ip');
+    my $tunnel_remote_ip = $self->get_ip(is_wicked_ref => 1, type => 'gre_tunnel_ip');
+    setup_tunnel(tunnel_ip => $tunnel_ip, config => $config, tunnel_iface => "gre1");
+    $results{1} = get_test_result($tunnel_remote_ip, "");
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: $results{1} = get_test_result($self, $tunnel_remote_ip, "");

File: tests/wicked/advanced_sut.pm [hunk 12]
```diff
@@ -21,44 +19,64 @@ use utils 'systemctl';
...
-    $ret = $self->ping_with_timeout(ip => "$parallel_ip_in_tunnel", timeout => '60', ip_version => 'v6');
+    #$self->before_scenario('Test 2', 'Create a gre interface from wicked XML files');
+
+    $self->before_scenario('Test 3', 'Create a SIT interface from legacy ifcfg files', $iface);
+    $config = '/etc/sysconfig/network/ifcfg-sit1';
+    $self->get_from_data('wicked/ifcfg-sit1_', $config, add_suffix => 1);
+    $tunnel_ip        = $self->get_ip(is_wicked_ref => 0, type => 'sit_tunnel_ip');
+    $tunnel_remote_ip = $self->get_ip(is_wicked_ref => 1, type => 'sit_tunnel_ip');
+    setup_tunnel(tunnel_ip => $tunnel_ip, config => $config, tunnel_iface => "sit1");
+    $results{3} = get_test_result($tunnel_remote_ip, "v6");
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: $results{3} = get_test_result($self, $tunnel_remote_ip, "v6");

File: tests/wicked/advanced_sut.pm [hunk 13]
```diff
@@ -21,44 +19,64 @@ use utils 'systemctl';
...
-    assert_script_run('ip a');
-    my $ret = $self->ping_with_timeout(ip => "$parallel_ip_in_tunnel", timeout => '60');
+    my %results;
+    my $iface  = script_output('ls /sys/class/net/ | grep -v lo | head -1');
+    my $config = '/etc/sysconfig/network/ifcfg-gre1';
+    $self->get_from_data('wicked/ifcfg-gre1_', $config, add_suffix => 1);
+    $self->before_scenario('Test 1', 'Create a gre interface from legacy ifcfg files');
+    my $tunnel_ip        = $self->get_ip(is_wicked_ref => 0, type => 'gre_tunnel_ip');
+    my $tunnel_remote_ip = $self->get_ip(is_wicked_ref => 1, type => 'gre_tunnel_ip');
+    setup_tunnel(tunnel_ip => $tunnel_ip, config => $config, tunnel_iface => "gre1");
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: setup_tunnel($self, tunnel_ip => $tunnel_ip, config => $config, tunnel_iface => "gre1");

File: tests/wicked/advanced_sut.pm [hunk 14]
```diff
@@ -21,44 +19,64 @@ use utils 'systemctl';
...
-    assert_script_run('ip a');
-    $ret = $self->ping_with_timeout(ip => "$parallel_ip_in_tunnel", timeout => '60', ip_version => 'v6');
+    #$self->before_scenario('Test 2', 'Create a gre interface from wicked XML files');
+
+    $self->before_scenario('Test 3', 'Create a SIT interface from legacy ifcfg files', $iface);
+    $config = '/etc/sysconfig/network/ifcfg-sit1';
+    $self->get_from_data('wicked/ifcfg-sit1_', $config, add_suffix => 1);
+    $tunnel_ip        = $self->get_ip(is_wicked_ref => 0, type => 'sit_tunnel_ip');
+    $tunnel_remote_ip = $self->get_ip(is_wicked_ref => 1, type => 'sit_tunnel_ip');
+    setup_tunnel(tunnel_ip => $tunnel_ip, config => $config, tunnel_iface => "sit1");
```
Comment (jlausuch) [inline tests/wicked/advanced_sut.pm]: setup_tunnel($self, tunnel_ip => $tunnel_ip, config => $config, tunnel_iface => "sit1");

**Comment by asmorodskyi** [comment]
@rtorrero , @mtomaschewski

**Pull request merged**